### PR TITLE
Handle social links with or without the link prefix

### PR DESCRIPTION
**Pull request by ghiculescu**
This should make more links work even if they were copy pasted rather than the user entering just their handle.

Fixes https://github.com/joemasilotti/railsdevs.com/issues/247

### Pull request checklist

- [x] Your code contains tests relevant for code you modified
- [x] You have linted and tested the project with `bin/check`

### CODE REVIEW

**Comment by joemasilotti** [comment]
Thanks for the PR! I like this as a temporary approach, but it doesn't address the underlying issue of folks not entering data the way the system expects. If you have the appetite, I'd prefer a different take.

### Validations

If we validate the content we should be able to enforce someone only typing in their username for each field. From some quick googling, platforms have the following restrictions:

* GitHub - alphanumeric with a single hyphen
* Twitter - alphanumeric with underscores
* LinkedIn - alphanumeric

From that, I think we can be liberal and create a `SocialUsernameValidator` that only accepts letters, numbers, hyphens, and underscores and use it for all 3. Then, the validation error can say "Your [GitHub|Twitter|LinkedIn] username can only contain letters, numbers, hyphens, and underscores."

I don't care about validating the single hyphen or that Twitter doesn't support hyphens, etc. Using the same validator for all 3 is fine. Especially because I don't want to have to update this if/when their policies change.

### Existing developer profiles

The tricky part is existing developer profiles. We have two options there:

1. Only run this validation if the previous value was blank. For example, this will trigger when creating a new profile and when adding a social username to an existing profile for the first time.
2. Writing a migration that converts all existing usernames to valid ones. This might be as simple as dropping the prefixes you listed above, but I can't say for sure without diving in to the data.

---

Let me know what you think and if you'd like to run with this!

**Comment by ghiculescu** [comment]
@joemasilotti thanks for the feedback. I think I respectfully disagree though. My view is that you should always be liberal in what input you accept and try to handle as many different inputs as possible if they can all lead to a reasonable output. (There's sometimes an exception if you want to report on the data and thus really need it to be uniform.)

So in this specific case, it seems like the root cause is some developers type in their handle "correctly" and some copy and paste the URL from their profile page. In both use cases, we can work out what outcome is expected by the user. Therefore I think we should support either method of input, rather than making the second user group jump through hoops to suit our preferred data format.

Basically, rather than saying this:

> it doesn't address the underlying issue of folks not entering data the way the system expects.

I think we should be asking, how can we make the system handle data entered the way the user expects it to work?

**Comment by joemasilotti** [comment]
Thanks for pushing back! After sleeping on this, I agree that we should accept as much as possible. However, I don't think we should store ill-formatted data in the database and rely on the presentation layer to clean it up.

My vote is to accept both usernames and full links in the form field. Then, using a version of your display logic, strip out the "prefix" and only store the username in the database. If extracted, we can use this to normalize all of the existing data, too.

I'm leaning towards an Active Record callback to trigger the normalization.

**Comment by joemasilotti** [comment]
Hey @ghiculescu, do you still have appetite to take this over the finish line? If not, no worries, I can pick up where you left off.

**Comment by joemasilotti** [comment]
All good! Thanks for the heads up. I'll take care of it.
